File: packages/android/core-domain/src/main/kotlin/app/notes/domain/Notifications.kt
package app.notes.domain

import java.time.Instant

/** Opaque id for a row in `public.reminders`. */
@JvmInline
value class ReminderId(val raw: String) {
    init { require(raw.isNotBlank()) { "ReminderId cannot be empty" } }
}

enum class ReminderStatus(val wireName: String) {
    Scheduled("scheduled"),
    Fired("fired"),
    Cancelled("cancelled"),
    ;

    companion object {
        fun fromWire(name: String): ReminderStatus =
            entries.firstOrNull { it.wireName == name }
                ?: throw IllegalArgumentException("Unknown ReminderStatus \"$name\"")
    }
}

data class Reminder(
    val id: ReminderId,
    val ownerId: UserId,
    val noteId: NoteId? = null,
    val fireAt: Instant,
    val title: String,
    val body: String? = null,
    val status: ReminderStatus = ReminderStatus.Scheduled,
) {
    init { require(title.isNotBlank()) { "Reminder.title cannot be empty" } }
}

data class PushPayload(
    val title: String,
    val body: String,
    val data: Map<String, String> = emptyMap(),
) {
    init {
        require(title.isNotBlank()) { "PushPayload.title cannot be empty" }
        require(body.isNotBlank()) { "PushPayload.body cannot be empty" }
    }
}

/**
 * The notifications port. V1 implementations bind locally
 * (AlarmManager on Android; setTimeout + Web Notifications on web).
 * V2 swaps in `FcmNotificationProvider` behind the same shape.
 */
interface NotificationProvider {
    suspend fun scheduleLocal(reminder: Reminder)
    suspend fun cancelLocal(reminderId: ReminderId)
    /** V2 only — V1 impls throw NotImplementedError. */
    suspend fun sendPush(userId: UserId, payload: PushPayload)
}

File: packages/domain/src/notifications.ts
import type { NoteId, UserId } from "./ids.ts";

/** Opaque id for a scheduled reminder row in `public.reminders`. */
export type ReminderId = string & { readonly __brand: "ReminderId" };

export function ReminderId(value: string): ReminderId {
  if (value.trim().length === 0) {
    throw new Error("ReminderId cannot be empty");
  }
  return value as ReminderId;
}

export type ReminderStatus = "scheduled" | "fired" | "cancelled";

export const REMINDER_STATUSES: readonly ReminderStatus[] = [
  "scheduled",
  "fired",
  "cancelled",
] as const;

/**
 * A scheduled reminder. V1 wires this to a local alarm (AlarmManager on
 * Android, `setTimeout` + Web Notifications API on web). V2's FCM swap
 * lights up `sendPush` through the same port.
 */
export interface Reminder {
  readonly id: ReminderId;
  readonly ownerId: UserId;
  /** Optional source note for the reminder body / deep link. */
  readonly noteId?: NoteId;
  readonly fireAt: string;
  readonly title: string;
  readonly body?: string;
  readonly status: ReminderStatus;
}

/** V2 push payload — placeholder shape until FCM lands. */
export interface PushPayload {
  readonly title: string;
  readonly body: string;
  readonly data?: Readonly<Record<string, string>>;
}

/**
 * The notifications port. V1 implementations:
 *  - Android: `AndroidNotificationProvider` (AlarmManager + BroadcastReceiver).
 *  - Web: `WebNotificationProvider` (`Notification` + `setTimeout` for <24h).
 * V2: a single `FcmNotificationProvider` replaces both at the binding
 * layer — domain code stays unchanged.
 */
export interface NotificationProvider {
  scheduleLocal(reminder: Reminder): Promise<void>;
  cancelLocal(reminderId: ReminderId): Promise<void>;
  /** V2 only — V1 impls reject with NotImplementedError. */
  sendPush(userId: UserId, payload: PushPayload): Promise<void>;
}
